metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java

```diff
@@ -180,18 +180,18 @@ private void extractBroaderResources(EnrichmentBase resource, Set<String> destin
     }
 
     // If not in the cache, or no vocabulary (ID) is known, we need to resolve the resource.
-    final VocabularyCandidates candidates =
-            VocabularyCandidates.findVocabulariesForUrl(resourceId, vocabularyDao::getByUriSearch);
     if (entityString == null || vocabulary == null) {
+      final VocabularyCandidates candidates =
+          VocabularyCandidates.findVocabulariesForUrl(resourceId, vocabularyDao::getByUriSearch);
       final Pair<String, Vocabulary> transformedEntity =
           retrieveTransformedEntity(resourceId, candidates);
       if (transformedEntity != null) {
         entityString = transformedEntity.getLeft();
         vocabulary = transformedEntity.getRight();
-        final ProcessedEntity entityToCache = new ProcessedEntity();
+        ProcessedEntity entityToCache = (cachedEntity == null) ? new ProcessedEntity() : cachedEntity;
         entityToCache.setXml(entityString);
         entityToCache.setResourceId(resourceId);
-        entityToCache.setVocabularyId(vocabulary == null ? null : vocabulary.getId());
+        entityToCache.setVocabularyId((vocabulary == null) ? null : vocabulary.getId().toString());
```

Comment:
I thought the id of Vocabulary was type ObjectId, it accepts String?

Reply:
The id of the vocabulary is ObjectId indeed but we save it as string in the ProcessedEntity..
A better way to do this is to use mongo references see here:
https://docs.mongodb.com/manual/reference/database-references/

But this was out of scope of this ticket, so I've kept the previous functionality here.